File: CulinarySiteApp/src/app/services/comment.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ICommentListModel } from '../interfaces/comment/comment-list-model.interface';
import { ICommentDetailModel } from '../interfaces/comment/comment-detail-model.interface';
import { CreateCommentModel } from '../viewmodels/comment/create-comment-model.class';
import { UpdateCommentModel } from '../viewmodels/comment/update-comment-model.class';

@Injectable()
export class CommentService {
  public url: string = '/api/comment';

  constructor(private http: HttpClient) {}

  public getCommentList(withRelated: boolean): Observable<ICommentListModel[]> {
    return this.http.get<ICommentListModel[]>(this.url + '/' + withRelated);
  }

  public getComment(
    id: number,
    withRelated: boolean
  ): Observable<ICommentDetailModel> {
    return this.http.get<ICommentDetailModel>(
      this.url + '/' + id + '/' + withRelated
    );
  }

  public createComment(
    createCommentModel: CreateCommentModel
  ): Observable<void> {
    return this.http.post<void>(this.url, createCommentModel);
  }

  public updateComment(
    updateCommentModel: UpdateCommentModel
  ): Observable<void> {
    return this.http.post<void>(this.url, updateCommentModel);
  }

  public deleteComment(id: number): Observable<void> {
    return this.http.delete<void>(this.url + '/' + id);
  }
}

File: CulinarySite/CulinarySite.Bll/Services/CommentService.cs
﻿using System.Collections.Generic;
using AutoMapper;
using CulinarySite.Bll.Interfaces;
using CulinarySite.Common.Dtos.Comment;
using CulinarySite.Dal.Interfaces;
using CulinarySite.Domain.Entities;
using System.Linq;

namespace CulinarySite.Bll.Services
{
    public class CommentService : ICommentService
    {
        private readonly IReadOnlyGenericRepository<Comment> _commentReadOnlyRepository;
        private readonly IWriteGenericRepository<Comment> _commentWriteRepository;
        private readonly IMapper _mapper;
        public CommentService(
            IReadOnlyGenericRepository<Comment> commentReadOnlyRepository,
            IWriteGenericRepository<Comment> commentWriteRepository,
            IMapper mapper)
        {
            _commentReadOnlyRepository = commentReadOnlyRepository;
            _commentWriteRepository = commentWriteRepository;
            _mapper = mapper;
        }

        public void CreateComment(CreateCommentDto createCommentDto)
        {
            Comment comment = _mapper.Map<Comment>(createCommentDto);

            _commentWriteRepository.Create(comment);
            _commentWriteRepository.Save();
        }
        public void UpdateComment(UpdateCommentDto updateCommentDto)
        {
            Comment comment = _mapper.Map<Comment>(updateCommentDto);

            _commentWriteRepository.Update(comment);
            _commentWriteRepository.Save();
        }

        public void DeleteComment(int id)
        {
            _commentWriteRepository.Delete(id);
            _commentWriteRepository.Save();
        }

        public CommentDetailDto GetComment(int id, bool withRelated)
        {
            var comment = new Comment();
            var commentDetailDto = new CommentDetailDto();

            if (withRelated)
            {
                comment = _commentReadOnlyRepository.GetItemWithInclude(
                x => x.Id == id,
                x => x.Restaurants
                );

                commentDetailDto = _mapper.Map<CommentDetailDto>(comment);

                return commentDetailDto;
            }

            comment = _commentReadOnlyRepository.GetItem(id);

            commentDetailDto = _mapper.Map<CommentDetailDto>(comment);

            return commentDetailDto;
        }

        public IEnumerable<CommentListDto> GetCommentList(bool withRelated)
        {
            IEnumerable<Comment> comments;
            IEnumerable < CommentListDto > commentListDtos;

            if (withRelated)
            {
                comments = _commentReadOnlyRepository.GetItemListWithInclude(x => x.Restaurants);

                commentListDtos = comments.Select(x => _mapper.Map<CommentListDto>(x));
                
                return commentListDtos;
            }

            comments = _commentReadOnlyRepository.GetItemList();

            commentListDtos = comments.Select(x => _mapper.Map<CommentListDto>(x));

            return commentListDtos;
        }
    }
}
